prov/sockets/src/sock_progress.c

```diff
@@ -2629,10 +2630,12 @@ static void sock_pe_poll(struct sock_pe *pe)
 	if (pe->waittime && ((fi_gettime_ms() - pe->waittime) < sock_pe_waittime))
 		return;
 
-	if (dlist_empty(&pe->tx_list) && dlist_empty(&pe->rx_list))
+	pthread_mutex_lock(&pe->list_lock);
+	if (dlist_empty(&pe->tx_list) && dlist_empty(&pe->rx_list)) {
+		pthread_mutex_unlock(&pe->list_lock);
 		return;
+	}
 
-	pthread_mutex_lock(&pe->list_lock);
 	if (!dlist_empty(&pe->tx_list)) {
```

Comment:
Something is still off here.  The above checks for dlist_empty, which is then followed by a check for !dlist_empty.  My guess is that the first if statement is a quick check without a lock to see if it even needs to do any work.  dlist_empty by itself just compares a pointer to itself really, so a lock isn't needed.  If the list is empty, the call returns without acquiring a lock.  If the list isn't empty, then the lock is acquired and the list is processed under lock (which includes checking if the list is empty).